model/task_queue_service_dependency.go

```diff
@@ -0,0 +1,319 @@
+package model
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/simple"
+	"gonum.org/v1/gonum/graph/topo"
+)
+
+type taskDistroDAGDispatchService struct {
+	mu          sync.RWMutex
+	distroID    string
+	graph       *simple.DirectedGraph
+	sorted      []graph.Node
+	itemNodeMap map[string]graph.Node
+	nodeItemMap map[int64]*TaskQueueItem
+	taskGroups  map[string]taskGroupTasks
+	ttl         time.Duration
+	lastUpdated time.Time
+}
+
+type taskGroupTasks struct {
+	id           string
+	runningHosts int // number of hosts task group is currently running on
+	maxHosts     int // number of hosts task group can run on
+	tasks        []TaskQueueItem
+}
+
+// taskDistroDAGDispatchService creates a taskDistroDAGDispatchService from a slice of TaskQueueItems.
+func newDistroTaskDAGDispatchService(distroID string, items []TaskQueueItem, ttl time.Duration) TaskDistroQueueService {
+	t := &taskDistroDAGDispatchService{
+		distroID: distroID,
+		ttl:      ttl,
+	}
+	t.graph = simple.NewDirectedGraph()
+	t.itemNodeMap = map[string]graph.Node{}
+	t.nodeItemMap = map[int64]*TaskQueueItem{}
+	t.taskGroups = map[string]taskGroupTasks{}
+
+	if len(items) != 0 {
+		t.rebuild(items)
+	}
+
+	return t
+}
+
+func (t *taskDistroDAGDispatchService) Refresh() error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if !shouldRefreshCached(t.ttl, t.lastUpdated) {
+		return nil
+	}
+
+	taskQueue, err := FindDistroTaskQueue(t.distroID)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	taskQueueItems := taskQueue.Queue
+	t.rebuild(taskQueueItems)
+
+	return nil
+}
+
+func (t *taskDistroDAGDispatchService) addItem(item TaskQueueItem) {
+	node := t.graph.NewNode()
+	t.graph.AddNode(node)
+	t.nodeItemMap[node.ID()] = &item
+	t.itemNodeMap[item.Id] = node
+}
+
+func (t *taskDistroDAGDispatchService) getItemByNodeID(id int64) *TaskQueueItem {
+	if item, ok := t.nodeItemMap[id]; ok {
+		return item
+	}
+	grip.Error(message.Fields{
+		"message": "programmer error, couldn't find node in map",
+	})
```

Comment:
shouldn't use message.Fields in this case, also probably shouldn't log an error here.